trillian/ctfe/handlers.go

```diff
@@ -441,94 +448,63 @@ func addPreChain(ctx context.Context, li *logInfo, w http.ResponseWriter, r *htt
 	return addChainInternal(ctx, li, w, r, true)
 }
 
-// GetTreeHead retrieves and builds a tree head structure for the given log; the returned
-// tree head is not yet signed.
-func GetTreeHead(ctx context.Context, client trillian.TrillianLogClient, logID int64, prefix string, remoteQuota *string) (*ct.SignedTreeHead, error) {
-	// Send request to the Log server.
-	req := trillian.GetLatestSignedLogRootRequest{LogId: logID}
-	if remoteQuota != nil {
-		req.ChargeTo = appendUserCharge(req.ChargeTo, *remoteQuota)
-	}
-	glog.V(2).Infof("%s: GetSTH => grpc.GetLatestSignedLogRoot %+v", prefix, req)
-	rsp, err := client.GetLatestSignedLogRoot(ctx, &req)
-	glog.V(2).Infof("%s: GetSTH <= grpc.GetLatestSignedLogRoot err=%v", prefix, err)
+// PingTreeHead retrieves a tree head for the given log, and updates the STH
+// timestamp metrics correspondingly.
+// TODO(pavelkalinnikov): Should we cache the resulting STH?
+func PingTreeHead(ctx context.Context, client trillian.TrillianLogClient, logID int64, prefix string) error {
+	slr, err := getSignedLogRoot(ctx, client, logID, prefix)
 	if err != nil {
-		return nil, err
-	}
-
-	// Check over the response.
-	slr := rsp.SignedLogRoot
-	if slr == nil {
-		return nil, errors.New("no log root returned")
-	}
-	glog.V(3).Infof("%s: GetSTH <= slr=%+v", prefix, slr)
-	if treeSize := slr.TreeSize; treeSize < 0 {
-		return nil, fmt.Errorf("bad tree size from backend: %d", treeSize)
-	}
-
-	if hashSize := len(slr.RootHash); hashSize != sha256.Size {
-		return nil, fmt.Errorf("bad hash size from backend expecting: %d got %d", sha256.Size, hashSize)
-	}
-
-	// Build the CT STH object, except the signature.
-	sth := ct.SignedTreeHead{
-		Version:   ct.V1,
-		TreeSize:  uint64(slr.TreeSize),
-		Timestamp: uint64(slr.TimestampNanos / 1000 / 1000),
+		return err
 	}
-	copy(sth.SHA256RootHash[:], slr.RootHash) // Checked size above.
-	lastSTHTimestamp.Set(float64(sth.Timestamp), strconv.FormatInt(logID, 10))
-	lastSTHTreeSize.Set(float64(sth.TreeSize), strconv.FormatInt(logID, 10))
-
-	return &sth, nil
+	lastSTHTimestamp.Set(float64(slr.TimestampNanos/1000/1000), strconv.FormatInt(logID, 10))
+	lastSTHTreeSize.Set(float64(slr.TreeSize), strconv.FormatInt(logID, 10))
+	return nil
 }
 
 func getSTH(ctx context.Context, li *logInfo, w http.ResponseWriter, r *http.Request) (int, error) {
-	var remoteQuotaUser *string
+	qctx := ctx
 	if li.instanceOpts.RemoteQuotaUser != nil {
 		rqu := li.instanceOpts.RemoteQuotaUser(r)
-		remoteQuotaUser = &rqu
+		qctx = context.WithValue(ctx, remoteQuotaCtxKey, rqu)
```

Comment:
withValue(qtcx ...) in case someone later adds another value before this.

Reply:
That's prudent.